scripts for ATAC-seq related analysis in the paper.
1. Encode ATAC-seq pipline (https://github.com/ENCODE-DCC/atac-seq-pipeline) for preproceesing.
2. makeATACTable.pl for summarize counts per regions.
3. DEseq2_ATAC.R for analyzing diffPeaks
4. run_mea_homer.sh for motif enrichment analysis
5. get_nearest_gene.py, chipseeker (use galaxy psu) and ABC enhancer-gene prediction (https://github.com/broadinstitute/ABC-Enhancer-Gene-Prediction) for annotate peaks.
6. run_deeptools.sh for normalize binding site activity.
